models/shopcart.py

```diff
@@ -1,35 +1,33 @@
+from .dataerror import DataValidationError
 class Shopcart(object):
-
     """
     This is model for the shop carts
     Assumptions:
         - In memory persistence
         - The fields of a shopcart include:
             - user id
-            - list of products (product id, quantity of product )
+            - Dictionary of products { product_id: quantity_of_product}
         - The total price of the shopcart will be dynamically calculated
-        - Model should only initially include `init` method and shopcart fields
-
     """
-
     __data = []
     __index = 0
 
-    __products = dict()
-
     def __init__(self, uid=0, products=None):
         """
         :param uid: user id
         :param products: dict of products <products id, quantity of product>
         """
-
         self.uid = int(uid)
-        self.products = products if products else []
+        self.products = self.__validate_products(products)
 
     def save(self):
         """ Saves a Shopcart in the database """
         Shopcart.__data.append(self)
 
+    def add_product(self, pq_tup):
```

Comment:
I think it would make sense to make this require `product` and `quantity` separate, with `quantity` defaulting to 1. This way you can form the structure you want to pass in to `validate_products` and the rest api doesn't have to know about the structure of our model.

Reply:
Done!